shapeshift/web
# fix: balance chart bugs

## Description

### fixes

* **correctly calculates eth balances by considering all erc20 txs, e.g. an airdrop claim is consider a fox receive but we weren't adjusting eth balance for the fee spent**
* **fixes portfolio chart by considering selected account type for bitcoin (this whole UX is jank AF)**
* **fixes timeframes for day and month to show a full day and month**

### improvements

* adjust timeframes for all timeframe options to be more accurate
* changes fiat balance in buckets to bignumber to support gigawhales with fed sized portfolios, and for safety/consistency
* add tests (mainly used to debug this, coverage isn't great)
### existing known bugs

* **still looking into parsing trades correctly** - DM me for wallet with ideal history
* **still not handling yearn vault assets correctly - waiting on market data for these**

### performance

* use binary search for asset price lookup rather than linear, before and afters below
```
    /*
    before
    usecalculatedBuckets, timeframe: 1Y}: 747.162109375 ms
    useBalanceChartData.ts:380 calculatedBuckets, timeframe: 1Y}: 174.860107421875 ms
    useBalanceChartData.ts:380 calculatedBuckets, timeframe: All}: 51.165771484375 ms
    useBalanceChartData.ts:380 calculatedBuckets, timeframe: All}: 266.466796875 ms
    */

   /*
    after
    calculatedBuckets, timeframe: 1Y}: 60.56298828125 ms
    useBalanceChartData.ts:380 calculatedBuckets, timeframe: 1Y}: 66.963134765625 ms
    useBalanceChartData.ts:380 calculatedBuckets, timeframe: All}: 39.866943359375 ms
    useBalanceChartData.ts:380 calculatedBuckets, timeframe: All}: 43.8037109375 ms
   */
```
## Notice

Before submitting a pull request, please make sure you have answered the following:


## Pull Request Type


## Issue (if applicable)

#462 

## Testing

Please outline all testing steps

test one

1. grab a wallet with a tx history less than 5 years from now
2. test the all chart
3. should display a zero balance at the left of the chart

test two

1. grab a wallet with btc and other assets
2. navigate to bitcoin and select an account type with no bitcoin
3. navigate back to dashboard and bitcoin shouldn't appear in balance chart
4. navigate to bitcoin and select an account type with bitcoin
5. navigate back to dashboard and bitcoin should appear in balance chart

test three

1. grab a wallet with a plain eth balance and erc20 txs not going back further than 5 years
2. check the eth asset balance chart, it should be zero at the beginning

## Screenshots (if applicable)

<img width="977" alt="Screen Shot 2021-11-22 at 5 06 12 PM" src="https://user-images.githubusercontent.com/88504456/142953092-9057ebc0-8244-4941-9534-c3a525a46602.png">
